diff --git a/charon/Cargo.toml b/charon/Cargo.toml
index 76e38f7e..33b47bc2 100644
--- a/charon/Cargo.toml
+++ b/charon/Cargo.toml
@@ -2,7 +2,7 @@
 name = "charon"
 version = "0.1.76"
 authors = ["Son Ho <hosonmarc@gmail.com>", "Guillaume Boisseau <nadrieril+git@gmail.com>"]
-edition = "2021"
+edition = "2024"
 license = "Apache-2.0"
 
 [lib]
diff --git a/charon/src/ids/vector.rs b/charon/src/ids/vector.rs
index dbc354e2..aecb95b2 100644
--- a/charon/src/ids/vector.rs
+++ b/charon/src/ids/vector.rs
@@ -260,7 +260,7 @@ where
         self.iter_indexed().map(|(id, _)| id)
     }
 
-    pub fn all_indices(&self) -> impl Iterator<Item = I> {
+    pub fn all_indices(&self) -> impl Iterator<Item = I> + use<I, T> {
         self.vector.indices()
     }
 
diff --git a/charon/src/transform/expand_associated_types.rs b/charon/src/transform/expand_associated_types.rs
index 28be9543..7cf2d870 100644
--- a/charon/src/transform/expand_associated_types.rs
+++ b/charon/src/transform/expand_associated_types.rs
@@ -794,7 +794,7 @@ impl<'a> ComputeItemModifications<'a> {
         &'b mut self,
         clause: &TraitClause,
         clause_to_path: fn(TraitClauseId) -> TraitRefPath,
-    ) -> impl Iterator<Item = AssocTypePath> + use<'b> {
+    ) -> impl Iterator<Item = AssocTypePath> + use<'a, 'b> {
         let trait_id = clause.trait_.skip_binder.trait_id;
         let clause_path = clause_to_path(clause.clause_id);
         self.compute_extra_params_for_trait(trait_id)
diff --git a/charon/src/transform/simplify_constants.rs b/charon/src/transform/simplify_constants.rs
index a0ba23eb..75b96d08 100644
--- a/charon/src/transform/simplify_constants.rs
+++ b/charon/src/transform/simplify_constants.rs
@@ -11,7 +11,7 @@
 //! handle the globals like function calls.
 
 use itertools::Itertools;
-use std::assert_matches::assert_matches;
+use std::assert_matches;
 
 use crate::transform::TransformCtx;
 use crate::ullbc_ast::*;
